kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+	}
+
+	return nil
+}
+
+// Wipe wipes the cache state.
+func (c *Cache) Wipe() {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	c.root = &cachedItem{
+		bucket: newCacheBucket(),
+	}
+}
+
+func (c *Cache) BeginReadTx() (walletdb.ReadTx, error) {
+	c.mx.RLock()
+
+	dbTx, err := c.backend.BeginReadTx()
+	if err != nil {
+		c.mx.RUnlock()
+		return nil, err
+	}
+
+	return newCacheReadTx(c, dbTx), nil
+}
+
+func (c *Cache) BeginReadWriteTx() (walletdb.ReadWriteTx, error) {
+	c.mx.Lock()
+
+	dbTx, err := c.backend.BeginReadWriteTx()
+	if err != nil {
+		c.mx.Unlock()
+		return nil, err
+	}
+
+	c.currRwTx = newCacheReadWriteTx(c, dbTx)
+	return c.currRwTx, nil
+}
+
+func (c *Cache) Copy(w io.Writer) error {
+	return fmt.Errorf("unavailable")
+}
+
+func (c *Cache) Close() error {
+	err := c.backend.Close()
+	c.Wipe()
+	return err
+}
+
+func (c *Cache) PrintStats() string {
+	return "<unavailable>"
+}
+
+func (c *Cache) View(f func(tx walletdb.ReadTx) error, reset func()) error {
+	tx, err := c.BeginReadTx()
+	if err != nil {
+		return err
+	}
+
+	reset()
+
+	err = f(tx)
+	rollbackErr := tx.Rollback()
+
+	if err != nil {
+		return err
+	}
+
+	return rollbackErr
+}
+
+func (c *Cache) Update(f func(tx walletdb.ReadWriteTx) error,
+	reset func()) error {
+
+	tx, err := c.BeginReadWriteTx()
+	if err != nil {
+		return err
+	}
+
+	reset()
+
+	// Apply the tx closure, rollback on error.
+	if err := f(tx); err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+
+	// Attempt to commit, rollback on error. Note that since we're the
+	// only committers Commit should only fail with database error and
+	// never with any error that we'd normally retry on.
+	if err := tx.Commit(); err != nil {
+		_ = tx.Rollback()
+		return err
+	}
+
+	return nil
+}
+
+func (c *Cache) Batch(f func(tx walletdb.ReadWriteTx) error) error {
+	return c.Update(f, func() {})
+}
+
+var _ RwTx = (*cacheReadWriteTx)(nil)
+
+type cacheReadTx struct {
+	cache  *Cache
+	dbTx   walletdb.ReadTx
+	active bool
+}
+
+func newCacheReadTx(cache *Cache, dbTx walletdb.ReadTx) *cacheReadTx {
+	return &cacheReadTx{
+		cache:  cache,
+		dbTx:   dbTx,
+		active: true,
+	}
+}
+
+func readBucketImpl(cache *Cache, dbTx walletdb.ReadTx,
+	key []byte) walletdb.ReadBucket {
+
+	if cacheItem := cache.root.bucket.get(key); cacheItem != nil {
+		return newCacheReadBucket(cache, cacheItem)
+	}
+
+	return nil
+}
+
+func forEachBucketImpl(cache *Cache, f func(key []byte) error) error {
+	cache.root.bucket.tree.Ascend(func(item btree.Item) bool {
+		c := item.(*cachedItem)
+		if f([]byte(c.key)) != nil {
+			return false
+		}
+
+		return true
+	})
+
+	return nil
+}
+
+func (c *cacheReadTx) ReadBucket(key []byte) walletdb.ReadBucket {
+	return readBucketImpl(c.cache, c.dbTx, key)
+}
+
+func (c *cacheReadTx) ForEachBucket(f func(key []byte) error) error {
+	return forEachBucketImpl(c.cache, f)
+}
+
+func (c *cacheReadTx) Rollback() error {
+	if c.active {
+		defer func() {
+			c.active = false
+			c.cache.mx.RUnlock()
```

Comment:
Ah ok I see the unlock here...

Reply:
Yes, so the idea is to hold the read/write lock during the whole txn.